src/nvim/api/events.c

```diff
@@ -0,0 +1,86 @@
+#include <assert.h>
+#include <inttypes.h>
+#include <stdbool.h>
+#include <stddef.h>
+#include <stdint.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include "klib/kvec.h"
+#include "nvim/api/events.h"
+#include "nvim/api/private/converter.h"
+#include "nvim/api/private/defs.h"
+#include "nvim/api/private/helpers.h"
+#include "nvim/api/private/validate.h"
+#include "nvim/api/ui.h"
+#include "nvim/assert_defs.h"
+#include "nvim/autocmd.h"
+#include "nvim/autocmd_defs.h"
+#include "nvim/channel.h"
+#include "nvim/channel_defs.h"
+#include "nvim/eval.h"
+#include "nvim/eval/typval.h"
+#include "nvim/event/defs.h"
+#include "nvim/event/loop.h"
+#include "nvim/event/multiqueue.h"
+#include "nvim/event/wstream.h"
+#include "nvim/globals.h"
+#include "nvim/grid.h"
+#include "nvim/highlight.h"
+#include "nvim/macros_defs.h"
+#include "nvim/main.h"
+#include "nvim/map_defs.h"
+#include "nvim/mbyte.h"
+#include "nvim/memory.h"
+#include "nvim/memory_defs.h"
+#include "nvim/msgpack_rpc/channel.h"
+#include "nvim/msgpack_rpc/channel_defs.h"
+#include "nvim/msgpack_rpc/packer.h"
+#include "nvim/msgpack_rpc/packer_defs.h"
+#include "nvim/option.h"
+#include "nvim/types_defs.h"
+#include "nvim/ui.h"
+
+/// Notification sent to the client channel in case of an error response to an async API request.
+///
+/// @param channel_id
+/// @param type Error type id as defined by `api_info().error_types`.
+/// @param msg Error message.
+void nvim_error_event(uint64_t channel_id, Integer type, String msg)
+  FUNC_API_REMOTE_ONLY
+{
+  // TODO(bfredl): consider printing message to user, as will be relevant
+  // if we fork nvim processes as async workers
+  ELOG("async error on channel %" PRId64 ": %s", channel_id, msg.size ? msg.data : "");
+}
+
+/// Tells Nvim when a host-terminal event occurred.
+///
+/// Supports these events:
+///
+///   - "termresponse": The host-terminal sent a DA1, OSC, DCS, or APC response sequence to Nvim.
+///                     The payload is the received response. Sets |v:termresponse| and fires
+///                     |TermResponse|.
+///
+/// @param channel_id
+/// @param event Event name
+/// @param value Event payload
+/// @param[out] err Error details, if any.
+void nvim_ui_term_event(uint64_t channel_id, String event, Object value, Error *err)
+  FUNC_API_SINCE(12) FUNC_API_REMOTE_ONLY
```

Comment:
This is an API, not an event.

Reply:
shit.  that is not supposed to happen, `_event` was reserved for events only.

Reply:
@zeertzjq what do you mean exactly? it is emitted as an event: https://github.com/neovim/neovim/blob/de950f827236e2c7447bd776c2509faa39c27b63/src/nvim/tui/input.c#L659

Reply:
> The TUI client sends all its API calls (`nvim_ui_attach`, `nvim_paste`, `nvim_input` etc.) as events.

Those APIs,

- do not have "notify" semantics (regardless of how a particular client invokes them)
- do not have return values
- are not suffixed with `_event`

I agree with your point about distinguishing Nvim-sourced vs client-sourced events.

Also, this discussion gets muddled if we decide to support [Lua + API in the TUI](https://github.com/neovim/neovim/pull/35406#issuecomment-3218420328)?